File: src/services/workspace.ts
import { invoke } from "@tauri-apps/api/core"
import type { Workspace, WorkspaceType } from "@/types"

interface WorkspaceRecord {
  id: string
  name: string
  icon?: string
  createdAt: string
  updatedAt: string
  agentConfig: { providerId: string; modelId: string; systemPrompt?: string }
  layout: { previewPanelWidth: number; activeRendererId?: string }
  workspaceType?: string
  repoPath?: string
  stateSummary?: string
  status: string
}

function fromRecord(r: WorkspaceRecord): Workspace {
  // Smart default: if repoPath is set and no explicit type, it must be linked
  const workspaceType: WorkspaceType =
    (r.workspaceType as WorkspaceType) ?? (r.repoPath ? "linked" : "internal")
  return {
    id: r.id,
    name: r.name,
    icon: r.icon,
    createdAt: r.createdAt,
    updatedAt: r.updatedAt,
    agentConfig: r.agentConfig,
    layout: r.layout,
    workspaceType,
    repoPath: r.repoPath,
    stateSummary: r.stateSummary,
    status: (r.status as Workspace["status"]) ?? "idle",
    lastActivity: r.updatedAt,
  }
}

function toRecord(ws: Workspace): WorkspaceRecord {
  return {
    id: ws.id,
    name: ws.name,
    icon: ws.icon,
    createdAt: ws.createdAt,
    updatedAt: ws.updatedAt,
    agentConfig: ws.agentConfig,
    layout: ws.layout,
    workspaceType: ws.workspaceType,
    repoPath: ws.repoPath,
    stateSummary: ws.stateSummary,
    status: ws.status,
  }
}

export async function listWorkspaces(): Promise<Workspace[]> {
  const records = await invoke<WorkspaceRecord[]>("list_workspaces")
  return records.map(fromRecord)
}

export async function createWorkspace(ws: Workspace): Promise<void> {
  await invoke("create_workspace", { record: toRecord(ws) })
}

export async function updateWorkspace(ws: Workspace): Promise<void> {
  await invoke("update_workspace", { record: toRecord(ws) })
}

export async function deleteWorkspace(id: string): Promise<void> {
  await invoke("delete_workspace", { id })
}

export function newWorkspace(
  name: string,
  providerId: string,
  modelId: string
): Workspace {
  const now = new Date().toISOString()
  return {
    id: crypto.randomUUID(),
    name,
    icon: "📁",
    createdAt: now,
    updatedAt: now,
    agentConfig: { providerId, modelId },
    layout: { previewPanelWidth: 0 },
    workspaceType: "internal",
    status: "idle",
  }
}

export function importWorkspace(
  folderPath: string,
  providerId: string,
  modelId: string
): Workspace {
  const parts = folderPath.split("/")
  const name = parts[parts.length - 1] || "Imported"
  return {
    ...newWorkspace(name, providerId, modelId),
    workspaceType: "linked",
    repoPath: folderPath,
    icon: "🔗",
  }
}

File: src-tauri/src/commands/workspace.rs
use crate::error::AppError;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceRecord {
    pub id: String,
    pub name: String,
    pub icon: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub agent_config: AgentConfig,
    pub layout: WorkspaceLayout,
    pub workspace_type: Option<String>, // "internal" | "linked"
    pub repo_path: Option<String>,
    pub state_summary: Option<String>,
    pub status: String, // "active" | "pending" | "idle"
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentConfig {
    pub provider_id: String,
    pub model_id: String,
    pub system_prompt: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceLayout {
    pub preview_panel_width: u32,
    pub active_renderer_id: Option<String>,
}

fn workspaces_dir() -> Result<PathBuf, AppError> {
    let home = dirs::home_dir()
        .ok_or_else(|| AppError::Other("Cannot resolve home dir".into()))?;
    Ok(home.join(".mindeck").join("workspaces"))
}

fn workspace_dir(base: &PathBuf, id: &str) -> PathBuf {
    base.join(id)
}

fn workspace_path(base: &PathBuf, id: &str) -> PathBuf {
    workspace_dir(base, id).join("workspace.json")
}

#[tauri::command]
pub fn list_workspaces() -> Result<Vec<WorkspaceRecord>, AppError> {
    let dir = workspaces_dir()?;
    if !dir.exists() {
        return Ok(vec![]);
    }
    let mut workspaces = Vec::new();
    for entry in fs::read_dir(&dir)? {
        let path = entry?.path();
        if !path.is_dir() {
            continue;
        }
        let meta_path = path.join("workspace.json");
        if !meta_path.exists() {
            continue;
        }
        let raw = fs::read_to_string(&meta_path)?;
        let record: WorkspaceRecord = serde_json::from_str(&raw)?;
        workspaces.push(record);
    }
    // Sort by updatedAt descending
    workspaces.sort_by(|a, b| b.updated_at.cmp(&a.updated_at));
    Ok(workspaces)
}

#[tauri::command]
pub fn create_workspace(record: WorkspaceRecord) -> Result<(), AppError> {
    let base = workspaces_dir()?;
    let ws_dir = workspace_dir(&base, &record.id);
    // Create workspace directory tree
    for sub in &["conversations", "knowledge/documents", "knowledge/index", "outputs", "files"] {
        fs::create_dir_all(ws_dir.join(sub))?;
    }
    let json = serde_json::to_string_pretty(&record)?;
    fs::write(workspace_path(&base, &record.id), json)?;
    Ok(())
}

#[tauri::command]
pub fn update_workspace(record: WorkspaceRecord) -> Result<(), AppError> {
    let base = workspaces_dir()?;
    let path = workspace_path(&base, &record.id);
    if !path.exists() {
        return Err(AppError::Other(format!("Workspace {} not found", record.id)));
    }
    let json = serde_json::to_string_pretty(&record)?;
    fs::write(path, json)?;
    Ok(())
}

#[tauri::command]
pub fn delete_workspace(id: String) -> Result<(), AppError> {
    let base = workspaces_dir()?;
    let ws_dir = workspace_dir(&base, &id);
    if ws_dir.exists() {
        fs::remove_dir_all(ws_dir)?;
    }
    Ok(())
}
